File: newspark-art/python/steps/pagination_steps.py
from behave import given, when, then
from pages.tasks_page import TasksPage

@given('there are multiple pages of tasks')
def step_impl(context):
    assert context.tasks_page.is_pagination_visible(), "Pagination not visible, no multiple pages of tasks"

@then('there is only one page of tasks')
def step_impl(context):
    assert not context.tasks_page.is_pagination_visible(), "Pagination is visible, there should be only one page of tasks"

@given('I am on a page other than the first page')
def step_impl(context):
    context.tasks_page.click_next()

@when('I click the Next button')
def step_impl(context):
    context.tasks_page.click_next()

@when('I click the Previous button')
def step_impl(context):
    context.tasks_page.click_previous()

@then('I should see the next page of tasks')
def step_impl(context):
    # Validate that the current page number has increased
    current_page = int(context.tasks_page.get_current_page_number())
    assert current_page > 1, f"Expected current page number to be greater than 1, but got {current_page}"

@then('I should see the first page of tasks')
def step_impl(context):
    # Validate that the current page number has decreased
    current_page = int(context.tasks_page.get_current_page_number())
    assert current_page == 0, f"Expected current page number to be 1, but got {current_page}"

@then('the current page number should be updated')
def step_impl(context):
    # Validate that the current page number is displayed correctly
    current_page = context.tasks_page.get_current_page_number()
    assert current_page.isdigit(), f"Expected current page number to be a digit, but got {current_page}"

@then('I should see the current page number')
def step_impl(context):
    assert context.tasks_page.get_current_page_number().isdigit(), "Current page number not visible"

@then('I should see the total number of pages')
def step_impl(context):
    assert context.tasks_page.get_total_pages().isdigit(), "Total number of pages not visible"

@then('the Next button should be disabled')
def step_impl(context):
    assert context.tasks_page.is_next_button_disabled(), "Next button is enabled, but it should be disabled"

@then('the Previous button should be disabled')
def step_impl(context):
    assert context.tasks_page.is_previous_button_disabled(), "Previous button is enabled, but it should be disabled"

File: newspark-art/CSharp/StepDefinitions/PaginationSteps.cs
using System;
using System.Threading.Tasks;
using Microsoft.Playwright;
using NUnit.Framework;
using Reqnroll;
using Newspark.PageObjects;

namespace Newspark.StepDefinitions;

[Binding]
public class PaginationSteps
{
    private readonly ScenarioContext _context;
    private readonly TasksPage _tasksPage;
    private readonly IPage _page;

    public PaginationSteps(ScenarioContext context)
    {
        _context = context;
        _page = (IPage)_context["Page"];
        _tasksPage = new TasksPage(_page);
    }

    [Given("I am on the tasks page")]
    [When("I am on the tasks page")]
    public async Task GivenIAmOnTheTasksPage()
    {
        await _tasksPage.Navigate();
    }

    [Given("there are multiple pages of tasks")]
    public async Task GivenThereAreMultiplePagesOfTasks()
    {
        Assert.IsTrue(await _tasksPage.IsPaginationVisible(), "Pagination not visible, no multiple pages of tasks");
    }

    [Then("there is only one page of tasks")]
    public async Task ThenThereIsOnlyOnePageOfTasks()
    {
        Assert.IsFalse(await _tasksPage.IsPaginationVisible(), "Pagination is visible, there should be only one page of tasks");
    }

    [Given("I am on a page other than the first page")]
    public async Task GivenIAmOnAPageOtherThanTheFirstPage()
    {
        await _tasksPage.ClickNext();
    }

    [When("I click the Next button")]
    public async Task WhenIClickTheNextButton()
    {
        await _tasksPage.ClickNext();
    }

    [When("I click the Previous button")]
    public async Task WhenIClickThePreviousButton()
    {
        await _tasksPage.ClickPrevious();
    }

    [Then("I should see the next page of tasks")]
    public async Task ThenIShouldSeeTheNextPageOfTasks()
    {
        int currentPage = await _tasksPage.GetCurrentPageNumber();
        Assert.Greater(currentPage, 1, $"Expected current page number to be greater than 1, but got {currentPage}");
    }

    [Then("I should see the first page of tasks")]
    public async Task ThenIShouldSeeTheFirstPageOfTasks()
    {
        int currentPage = await _tasksPage.GetCurrentPageNumber();
        Assert.AreEqual(1, currentPage, $"Expected current page number to be 1, but got {currentPage}");
    }

    [Then("the current page number should be updated")]
    public async Task ThenTheCurrentPageNumberShouldBeUpdated()
    {
        string currentPage = await _tasksPage.GetCurrentPageNumberText();
        Assert.IsTrue(int.TryParse(currentPage, out _), $"Expected current page number to be a digit, but got {currentPage}");
    }

    [Then("I should see the current page number")]
    public async Task ThenIShouldSeeTheCurrentPageNumber()
    {
        string currentPage = await _tasksPage.GetCurrentPageNumberText();
        Assert.IsTrue(int.TryParse(currentPage, out _), "Current page number not visible");
    }

    [Then("I should see the total number of pages")]
    public async Task ThenIShouldSeeTheTotalNumberOfPages()
    {
        string totalPages = await _tasksPage.GetTotalPages();
        Assert.IsTrue(int.TryParse(totalPages, out _), "Total number of pages not visible");
    }

    [Then("the Next button should be disabled")]
    public async Task ThenTheNextButtonShouldBeDisabled()
    {
        Assert.IsTrue(await _tasksPage.IsNextButtonDisabled(), "Next button is enabled, but it should be disabled");
    }

    [Then("the Previous button should be disabled")]
    public async Task ThenThePreviousButtonShouldBeDisabled()
    {
        Assert.IsTrue(await _tasksPage.IsPreviousButtonDisabled(), "Previous button is enabled, but it should be disabled");
    }
}
